### PR TITLE
Azure sdk upgrade from v59.2.0 to v60.0.0

### PR DESCRIPTION
**Pull request by xiaxyi**
Upgrading azure-go-sdk from 59.2.0 to 60.0.0  in order to use the stable sdk  (2021-10-01) of the web pubsub service.

Changed data factory resource because:
PR [#16563](https://github.com/Azure/azure-rest-api-specs/pull/16563/files) DataFactory removed some sub-classes of the property `datasetCompressionType` and this resulted in the missing of the enums in the sdk. 

Considering  that the service side won't use those removed sub-classes anymore, instead, use the base-class in the future so I removed the corresponding enums from TF as well. 

ACCTest:
```
$ make acctests SERVICE='datafactory' TESTARGS=' -run=TestAccDataFactoryDatasetBinary_blob' TESTTIMEOUT='60m'
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/datafactory   391.610s

$ make acctests SERVICE='datafactory' TESTARGS=' -run=TestAccDataFactoryDatasetBinary_blob_with_filepath' 
--- PASS: TestAccDataFactoryDatasetBinary_blob_with_filepath (386.66s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/datafactory   387.539s

$ make acctests SERVICE='datafactory' TESTARGS=' -run=TestAccDataFactoryDatasetBinary_http' TESTTIMEOUT='60m'
--- PASS: TestAccDataFactoryDatasetBinary_http (295.98s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/datafactory   296.728s

$ make acctests SERVICE='datafactory' TESTARGS=' -run=TestAccDataFactoryDatasetBinary_sftp' TESTTIMEOUT='60m'
--- PASS: TestAccDataFactoryDatasetBinary_sftp (366.17s)
--- PASS: TestAccDataFactoryDatasetBinary_sftpComplete (366.51s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/datafactory   367.401s

$ make acctests SERVICE='datafactory' TESTARGS=' -run=TestAccDataFactoryDatasetBinary_sftpComplete' 
--- PASS: TestAccDataFactoryDatasetBinary_sftpComplete (308.17s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/datafactory   309.229s
```

### CODE REVIEW

File: internal/services/datafactory/data_factory_dataset_binary_resource.go [hunk 1]
```diff
@@ -229,16 +230,21 @@ func resourceDataFactoryDatasetBinary() *pluginsdk.Resource {
 						},
 						// SFTP Specific field
 						"type": {
-							Type:     pluginsdk.TypeString,
-							Required: true,
+							Type:             pluginsdk.TypeString,
+							Required:         true,
+							DiffSuppressFunc: suppress.CaseDifference,
```
Comment (tombuildsstuff) [inline internal/services/datafactory/data_factory_dataset_binary_resource.go:235]: this indicates a bug in the API that needs to be resolved/worked around above - so we need to do that and remove this here:

```suggestion
```

File: internal/services/datafactory/data_factory.go [hunk 2]
```diff
@@ -520,79 +520,36 @@ func flattenDataFactoryDatasetSFTPLocation(input *datafactory.SftpLocation) []in
...
-		result["level"] = compression.Level
-	}
-	if compression, ok := input.AsDatasetZipDeflateCompression(); ok {
-		result["type"] = compression.Type
-		result["level"] = compression.Level
+	result["type"] = input.Type
+	var level string
+
+	if input.Level != nil {
+		level = input.Level.(string)
```
Comment (tombuildsstuff) [inline internal/services/datafactory/data_factory.go:533]: this'd also work:

```suggestion
		level = string(input.Level)
```

File: internal/services/datafactory/data_factory_dataset_binary_resource.go [hunk 3]
```diff
@@ -229,16 +230,21 @@ func resourceDataFactoryDatasetBinary() *pluginsdk.Resource {
...
-								string(datafactory.TypeBasicDatasetCompressionTypeTar),
-								string(datafactory.TypeBasicDatasetCompressionTypeTarGZip),
-								string(datafactory.TypeBasicDatasetCompressionTypeZipDeflate),
-							}, false),
+								string(datafactory.CompressionCodecBzip2),
+								string(datafactory.CompressionCodecDeflate),
+								string(datafactory.CompressionCodecGzip),
+								string(datafactory.CompressionCodecLz4),
+								string(datafactory.CompressionCodecLzo),
+								string(datafactory.CompressionCodecNone),
```
Comment (tombuildsstuff) [inline internal/services/datafactory/data_factory_dataset_binary_resource.go:242]: this generally isn't a valid value.. can we either prove this value works or remove this

```suggestion
```

File: internal/services/datafactory/data_factory_dataset_binary_resource.go [hunk 4]
```diff
@@ -229,16 +230,21 @@ func resourceDataFactoryDatasetBinary() *pluginsdk.Resource {
...
+								string(datafactory.CompressionCodecSnappy),
+								string(datafactory.CompressionCodecTar),
+								string(datafactory.CompressionCodecTarGZip),
+								string(datafactory.CompressionCodecZipDeflate),
+							}, true),
```
Comment (tombuildsstuff) [inline internal/services/datafactory/data_factory_dataset_binary_resource.go:247]: (as above) this should be case sensitive here

```suggestion
							}, false),
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @xiaxyi 

Thanks for this PR.

Taking a look through it appears there's a couple of issues we need to workaround here - so we'll need to work those out to be able to proceed here.

Thanks!

File: internal/services/datafactory/data_factory_dataset_binary_resource.go (hunk 4 above)
Comment (xiaxyi) [reply internal/services/datafactory/data_factory_dataset_binary_resource.go:247]: Thanks @tombuildsstuff for the suggestion! Data factory removed the sub-classes of the dataset compression type as listed below and this led to the missing of the enums:
```
Removed:
const (
	TypeBasicDatasetCompressionTypeBZip2 TypeBasicDatasetCompression = "BZip2"
	TypeBasicDatasetCompressionTypeDatasetCompression TypeBasicDatasetCompression = "DatasetCompression"
	TypeBasicDatasetCompressionTypeDeflate TypeBasicDatasetCompression = "Deflate"
	TypeBasicDatasetCompressionTypeGZip TypeBasicDatasetCompression = "GZip"
	TypeBasicDatasetCompressionTypeTar TypeBasicDatasetCompression = "Tar"
	TypeBasicDatasetCompressionTypeTarGZip TypeBasicDatasetCompression = "TarGZip"
	TypeBasicDatasetCompressionTypeZipDeflate TypeBasicDatasetCompression = "ZipDeflate"
)
```
Considering the actual supported value is the same, so I chose to use below enums:
```
const (
	CompressionCodecBzip2 CompressionCodec = "bzip2"
	CompressionCodecDeflate CompressionCodec = "deflate"
	CompressionCodecGzip CompressionCodec = "gzip"
	CompressionCodecLz4 CompressionCodec = "lz4"
	CompressionCodecLzo CompressionCodec = "lzo"
	CompressionCodecNone CompressionCodec = "none"
	CompressionCodecSnappy CompressionCodec = "snappy"
	CompressionCodecTar CompressionCodec = "tar"
	CompressionCodecTarGZip CompressionCodec = "tarGZip"
	CompressionCodecZipDeflate CompressionCodec = "zipDeflate"
)
```

hope it make sense to you, feel free to let me know if you have any questions

File: internal/services/datafactory/data_factory_dataset_binary_resource.go (hunk 3 above)
Comment (xiaxyi) [reply internal/services/datafactory/data_factory_dataset_binary_resource.go:242]: Thanks, I will remove it

File: internal/services/datafactory/data_factory_dataset_binary_resource.go (hunk 1 above)
Comment (xiaxyi) [reply internal/services/datafactory/data_factory_dataset_binary_resource.go:235]: Both "Gzip" and "gzip" is working, I think it's safe to remove this func

File: internal/services/datafactory/data_factory.go (hunk 2 above)
Comment (xiaxyi) [reply internal/services/datafactory/data_factory.go:533]: I'm afraid String() is not gonna work since the input value is an interface type.

**Comment by xiaxyi** [comment]
@tombuildsstuff Thanks for the suggestion, I added my feedbacks and modified some of the codes. feel free to let me know if you have any other concerns. Thanks.

File: internal/services/datafactory/data_factory_dataset_binary_resource.go (hunk 4 above)
Comment (tombuildsstuff) [reply internal/services/datafactory/data_factory_dataset_binary_resource.go:244]: @xiaxyi fwiw Golang uses Constants - not Enums, they work differently.

In this case can we fix the casing in those constants to be consistent with the original values - recasing these causes the issues highlighted above

Comment (xiaxyi) [reply internal/services/datafactory/data_factory_dataset_binary_resource.go:244]: I think we can do that, at lease this doesn't affects the case sensitive scenario. I'll modify it and run acctest. besides this, do you ok with the expand/flatten part?

**Comment by xiaxyi** [comment]
will modify the enums and try again

**Comment by xiaxyi** [comment]
@tombuildsstuff Thanks for the suggestion, I changed the code per the suggestion, can you help to check again? Thanks!

**Comment by tombuildsstuff** [comment]
hey @xiaxyi 

Thanks for pushing those changes - I hope you don't mind but since we need to ensure the casing is correct here I've opened #14667 which will make those changes and upgrade us to v60.0.0 of the Azure SDK for Go.

Thanks!

**Pull request closed without merging**